Sprint 3/step 1

sadova-nadezhda opened the pull request.

sadova-nadezhda committed:
update project

базовая настройка

sadova-nadezhda committed:
update project

- Создала страницы авторизации и регистрации
- Настроила переходы
- Создала страницы восстановления и сброса пароля

sadova-nadezhda committed:
update project

POST запрос к эндпоинтам

sadova-nadezhda committed:
update

create mkdir actions slice for auth
added Ts for Forms

sadova-nadezhda committed:
update project

create profile page

sadova-nadezhda committed:
update project

added slice and actions for auth
update Forms

sadova-nadezhda committed:
Update actions.ts

added createAsyncThunk

sadova-nadezhda committed:
update project

sadova-nadezhda committed:
update project

create IngredientPage
updated ProfilePage/ProfileForm

sadova-nadezhda committed:
update OrderDetails

added resetAllIngredientCounts

sadova-nadezhda committed:
update project

update AppHeader

Пытаюсь реализовать Защищённые маршруты в приложении

sadova-nadezhda committed:
update

откат auth/actions

sadova-nadezhda committed:
Update AppHeader.tsx

sadova-nadezhda committed:
update

added useNavigate

sadova-nadezhda committed:
update

Защищённые маршруты в приложении

sadova-nadezhda committed:
update project

sadova-nadezhda committed:
update

функционал /ingredients/:id

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации


* Не открывается попап с деталями ингредиента
* По заданию нужно сделать логику, при которой бы показывался попап с деталями ингредиента, если кликнуть по ингредиенту, но если скопировать ссылку на ингредиент при открытом попапе и **вставить в окно браузера**, то должна уже показываться **страница** с этими деталями. 
* Вам нужно использовать средства **роутинга** для определения, показывать ли попап с деталями, или показывать страницу. Вам должны были дать ссылку на документацию по этой логике. Вот она https://v5.reactrouter.com/web/example/modal-gallery . Пожалуйста, посоветуйтесь с наставником. В 6й версии роутинга примерно такая же логика.
* Если **перезагрузить** страницу с открытым попапом ингредиента, то этот открытый попап опять должен появляться (или страница с деталями). А сейчас ничего не отображается после перезагрузки.  Модальное окно теперь должно быть только в `App` во **2м `Routes`**, чтобы сработала логика с перезагрузкой страницы. Оно будет появляться только при наличии `background`

sadova-nadezhda committed:
update project

update Modal
update handleIngredientClick

sadova-nadezhda committed:
update project

обновила логику модального окна у ингредиентов

sadova-nadezhda committed:
update project

Обновила модалки
Оформить заказ может только авторизованный пользователь, иначе отправляет на /login

sadova-nadezhda committed:
Update IngredientModal.tsx

sadova-nadezhda committed:
Update IngredientDetails.tsx

sadova-nadezhda committed:
update project

обновила логику стр Профиля

sadova-nadezhda committed:
update project

обновила auth/actions

sadova-nadezhda committed:
update styles form

gennady-bars commented on src/services/auth/actions.ts:
```diff
@@ -0,0 +1,269 @@
+
+      const data = await apiRequest(`${BASE_URL}/auth/user`, 'GET', undefined, dispatch);
+
+      if (data.success) {
```
**Можно лучше**


Если интересно, то можно создать **универсальную** функцию запроса `request` , которая будет содержать в себе уже **базовый урл**, проверку на **ok** и проверку на `success`, чтобы не дублировать их в каждом запросе.

```js

// 1 раз объявляем базовый урл
export const BASE_URL = "https://norma.nomoreparties.space/api/";

// создаем функцию проверки ответа на `ok`
const checkResponse = (res) => {
  if (res.ok) {
    return res.json();
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ошибка ${res.status}`);
};

// создаем функцию проверки на `success`
const checkSuccess = (res) => {
  if (res && res.success) {
    return res;
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ответ не success: ${res}`);
};

// создаем универсальную фукнцию запроса с проверкой ответа и `success`
// В вызов приходит `endpoint`(часть урла, которая идет после базового) и опции
const request = (endpoint, options) => {
  // а также в ней базовый урл сразу прописывается, чтобы не дублировать в каждом запросе
  return fetch(`${BASE_URL}${endpoint}`, options)
    .then(checkResponse)
    .then(checkSuccess);
};

```


И теперь не нужно будет дублировать все это в каждом запросе. Нужно просто заменить в запросах все `fetch` на `request`, удалив лишние проверки и базовый урл. Все остальное останется без изменений

Код станет намного чище. Будет мелкий аналог `axios`

Пример запроса ингредиентов

```js
const getIngredients =  () => request("ingredients");

```

В `get`-запросах даже не нужно указывать 2й параметр. Он получится `undefined`, что без проблем для `fetch` (который внутри).

gennady-bars commented on src/components/Form/LoginForm.tsx:
```diff
@@ -0,0 +1,51 @@
+  const navigate = useNavigate();
+  const { loading, error } = useAppSelector((state) => state.auth);
+
+  const [email, setEmail] = useState<string>('');
+  const [password, setPassword] = useState<string>('');
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars commented on src/components/Form/LoginForm.tsx:
```diff
@@ -0,0 +1,51 @@
+  };
+
+  return (
+    <form onSubmit={onSubmit} className="mb-20">
```
Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. Обработка события `submit` еще срабатывает при нажатии на `Enter`, так как это встроенная логика сабмитов в браузере.

gennady-bars commented on src/components/Modal/IngredientModal.tsx:
```diff
@@ -0,0 +1,41 @@
+
+  useEffect(() => {
+      if (!ingredients.length) {
+        dispatch(fetchIngredients());
```
`dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/pages/IngredientPage/IngredientPage.tsx:
```diff
@@ -0,0 +1,46 @@
+
+  useEffect(() => {
+    if (!ingredients.length) {
+      dispatch(fetchIngredients());
```
`dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/components/Modal/Modal.tsx:
```diff
@@ -1,53 +1,57 @@
-export default function Modal({ title, onClose, children } : ModalProps) {
-  
+export default function Modal({ title, onClose, children }: ModalProps) {
+  const location = useLocation();
+    const { isModalOpen } = useOrderModal();
```
`useOrderModal` не нужно использовать внутри `Modal`, так как не имеет эффекта. Модальное окно открывается **снаружи**, а не врутри

gennady-bars commented on src/components/Modal/Modal.tsx:
```diff
@@ -1,53 +1,57 @@
-  const handleModalClick = (e: MouseEvent) => {
-    e.stopPropagation();
-  };
+  useEffect(() => {
+    if (!location.state?.background && isModalOpen) {
+      onClose();
+    }
+  }, [location, onClose, isModalOpen]);
```
Это нужно удалить

`useOrderModal` не нужно использовать внутри `Modal`, так как не имеет эффекта. Модальное окно открывается **снаружи**, а не врутри

gennady-bars commented on src/components/App/App.tsx:
```diff
@@ -1,33 +1,78 @@
 
 function App() {
+  const location = useLocation();
+  const background = location.state?.background;
```
Отлично сделана логика отображения попапа и страницы  под ним

gennady-bars commented on src/components/App/ProtectedRouteElement.tsx:
```diff
@@ -0,0 +1,23 @@
+  fallBackRoute: string;
+}
+
+const ProtectedRouteElement = ({ children, fallBackRoute }: ProtectedRouteElementProps) => {
+  const isAuthenticated = useAppSelector((state) => !!state.auth.user);
+  const navigate = useNavigate();
+
+  useEffect(() => {
+    if (!isAuthenticated) {
+      navigate(fallBackRoute);
+    }
+  }, [isAuthenticated, navigate, fallBackRoute]);
+
+  return <>{children}</>;
```
**Можно лучше**

Если интересно, то можно сделать вот так универсальный компонент `ProtectedRoute` с переадресацией

```jsx


export default function ProtectedRoute({ children, anonymous = false }) {
  const isLoggedIn = useSelector((store) => store.user.isLoggedIn);

  const location = useLocation();
  const from = location.state?.from || '/';
  // Если разрешен неавторизованный доступ, а пользователь авторизован...
  if (anonymous && isLoggedIn) {
    // ...то отправляем его на предыдущую страницу
    return <Navigate to={ from } />;
  }

  // Если требуется авторизация, а пользователь не авторизован...
  if (!anonymous && !isLoggedIn) {
    // ...то отправляем его на страницу логин
    return <Navigate to="/login" state={{ from: location}}/>;
  }

  // Если все ок, то рендерим внутреннее содержимое
  return children;
}

```

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:




* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку.
* Отлично сделана логика отображения попапа и страницы  под ним

но есть некоторые недочеты:


* Если пользователь **не авторизован** и пытается попасть на защищённый маршрут — его **переадресовывает** на маршрут `/login`. После авторизации пользователь автоматически переадресовывается на тот маршрут, к которому у него **не получилось** получить доступ ранее.  
* Нет **автологина**. Если перезагрузить страницу в профиле, то я опять оказываюсь на странице логина, а должен быть автологин и я должен попадать туда, откуда перезагружал страницу
* Не работает пункт меню `Конструктор`  в хэдере из профиля https://disk.yandex.ru/i/UHVTs-7dP3tYXA . Это тогда, когда я делал заказ, а меня перекинуло на логин, а после логина в профиль (хотя нужно к заказу обратно меня перекидывать)
* После успешного оформления заказа не работает пункт меню `Личный кабинет` в хэдере https://disk.yandex.ru/i/m5Azhy4OR8rvjQ 
* `dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.
* `useOrderModal` не нужно использовать внутри `Modal`, так как не имеет эффекта. Модальное окно открывается **снаружи**, а не врутри


**МОЖНО ЛУЧШЕ**


* при клике на **логотип** обычно идет переход на **главную** страницу

sadova-nadezhda committed:
Update AppHeader.tsx

при клике на логотип идет переход на главную страницу

sadova-nadezhda committed:
update project

Если пользователь не авторизован и пытается попасть на защищённый маршрут — его переадресовывает на маршрут /login. После авторизации пользователь автоматически переадресовывается на тот маршрут, к которому у него не получилось получить доступ ранее.

sadova-nadezhda committed:
update project

добавила автологин
Если перезагрузить страницу в профиле, то там и остаешься

sadova-nadezhda committed:
update project

Модальное окно открывается снаружи

sadova-nadezhda committed:
update project

dispatch(fetchIngredients()) 1 раз вызывается в App

sadova-nadezhda committed:
update project

универсальный кастомный хук для контроля любого количества инпутов в любых формах

sadova-nadezhda committed:
Update LoginForm.tsx

sadova-nadezhda committed:
update ProtectedRoute

универсальный компонент ProtectedRoute

sadova-nadezhda committed:
update Route path="/reset-password"

неавторизованный пользователь не может напрямую попасть на маршрут /reset-password

sadova-nadezhda commented:
После успешного оформления заказа не работает пункт меню Личный кабинет в хэдере
- у меня все работает

gennady-bars reviewed (commented):
Уже почти все отлично, но:

* Нет **автологина**. Если перезагрузить страницу в профиле, то я опять оказываюсь на странице логина, а должен быть автологин и я должен попадать туда, откуда перезагружал страницу
* `dispatch(fetchIngredients());` нужно удалить из `BurgerIngredients`. Все делается теперь в `App`

sadova-nadezhda committed:
update project

удалила лишний dispatch(fetchIngredients());

sadova-nadezhda committed:
update project

- добавила вывод ошибки
- исправила автологин (его реализовала в прошлой ссылке, но из-за того что обновила ProtectedRoute для улучшения, слетел автологин )
чуть-чуть оптимизировала auth/actions

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
